### ISSUE TITLE
Not finding linux pact mock service on centos 7

### ISSUE THREAD

**Issue by soundstep** [pact-foundation/pact-node#42]
I was previously using nodejs 4.3.2 on a centos 7 linux box running jenkins.
Since I've updated to nodejs 8.3.0, I can't seem to be able to require `@pact-foundation/pact-node`.
I get the following error:

```
Warning: Cannot find module '@pact-foundation/pact-mock-service-linux-x64' Use --force to continue.
```

I tried with `@pact-foundation/pact-node` version v4.8.1 and v4.12.0 (latest).

I first noticed the problem with karma-pact, but the problem happens in another repo where I don't use karma-pact.

**Comment by soundstep** [comment]
Update, not related to nodejs v8.3.0 (unless it doesn't happen all the time).
I'm now able to load pact-node when I revert it to v4.8.1, so maybe just a problem with this v4.12.0.
Hope that helps.

**Comment by soundstep** [comment]
Sorry another update, really hard to track what's going on.
I had the error once with nodejs v8.3.0 and and pact-node v4.8.1, so it is related to the latest nodejs version as I've never had it before, just not happening all the time.

The failure happens in a grunt task that looks like in case it is relevant:

```
grunt.registerTask('pact-service', 'Start pact mock service', function () {
        var done = this.async();
        var path = require('path');
        var pactWrapper = require('@pact-foundation/pact-node');
```

**Comment by mefellows** [comment]
As you may have guessed, we use optional dependencies to pull down dynamically (install time) the correct OS-specific package, in this case linux 64 bit. It's common to see that warning, but of course if it doesn't install one of the packages things won't work!

Are you able to please enable detailed logging for the npm install step (forgotten how off the top of my head) and share a gist of the output? That might help us track it down.

FWIW we currently test Pact Nnode with whatever TravisCI provides, currently `v8.2.1`.

**Comment by soundstep** [comment]
OS: Linux Centos 7
Nodejs: v8.3.0
NPM: 5.3.0
Pact-node: 4.8.1

You can find 2 log files in this gist, a npm install verbose and the jenkins job where the following error appears: 

```
Error: Cannot find module '@pact-foundation/pact-mock-service-linux-x64'
```

https://gist.github.com/soundstep/e8c1b72707fa18a041e591b7ed7b543b

Hope that helps!

**Comment by mboudreau** [comment]
This is probably because if the way pact node tries to discover the os
specific version. I should have time to fix this tomorrow and try out node
8.

**Comment by soundstep** [comment]
Hi @mboudreau 
Funny thing is that sometimes it is working, it is 50/50 I would say in our pipeline. There must be some timings involved? Did you see anything wrong in these logs?
Let me know if you guys can reproduce.

**Comment by mboudreau** [comment]
I'll look into it and report back.  There shouldn't be any "timing"
involved with the dependency, but there is a timeout in case the executable
doesn't spin up.  It would be helpful if you are to produce a dockerfile
that means me reproduce the issue.  Thanks.

**Comment by soundstep** [comment]
I'll make to make some time to try to reproduce the problem with a simple example. Meanwhile if you find the problem, let me know ;)

**Comment by soundstep** [comment]
Another update, I reproduced the problem on this Centos 7 instance fairly easily with these files:
https://gist.github.com/soundstep/8989b6c1c431458c8e4da3a9d0bfe8ba

I tried to dockerize it, but I can't reproduce the problem at all.
Seems to be related to speed, timing, execute timing, race condition or something similar?

The linux instance is a 4 CPU core (`Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz`), with 8BG of RAM, so pretty beefy.

I'm going to have a go on other Centos 7 instances, and in vagrant on my local machine.

Can you give some details on how you do the os detection? Maybe a small script that I can try out?

```
CentOS Linux release 7.3.1611 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.3.1611 (Core)
CentOS Linux release 7.3.1611 (Core)
```

**Comment by soundstep** [comment]
Here is the error it generates BTW:

```module.js:491
    throw err;
    ^

Error: Cannot find module '@pact-foundation/pact-mock-service-linux-x64'
    at Function.Module._resolveFilename (module.js:489:15)
    at Function.Module._load (module.js:439:25)
    at Module.require (module.js:517:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/romuquan/test-centos-pact/node_modules/@pact-foundation/pact-mock-service/index.js:7:18)
    at Module._compile (module.js:573:30)
    at Object.Module._extensions..js (module.js:584:10)
    at Module.load (module.js:507:32)
    at tryModuleLoad (module.js:470:12)
    at Function.Module._load (module.js:462:3)```

**Comment by mefellows** [comment]
Thanks @soundstep - I wish all bug reports came with as much info! This should enable us to get to the bottom of it.

**Comment by mboudreau** [comment]
@soundstep you are an absolute champ! Looking at the problem now.

**Comment by mboudreau** [comment]
@soundstep I'm 90% sure this has nothing to do with node, but instead npm. As part of the install step, it tries to install an OS specific binary of the pact mock service.  By default, it tries to download all of them, but npm will ignore all of them except for one because of the package restrictions (must be on "linux" and have architecture "x64").  I'm guessing that something might have changed in npm that makes it think that centos is not linux or that the architecture is wrong.  I'm digging in further by using a centos image on my virtualbox since it seems to work in docker (weird!) and can't reproduce.

**Comment by mboudreau** [comment]
@soundstep erm, I can't seem to be able to reproduce your issue using the centos7 image in Virtualbox.  Every time I run node, it seems to find the pact-mock-service correctly.  However, I did notice that in your gist, you seem to be ignoring the `npm install` step (as per what is in your output), which is quite important or else it won't get the OS specific binaries needed.

**Comment by soundstep** [comment]
Npm install is not there? I did test manually though. I will re-check today
and keep you posted.

Your guess sounds right about npm, did you notice anything special in the
verbose log in the first gist?

**Comment by mboudreau** [comment]
From your gist:
```
[vagrant@ie11win7 app]$ node -v
v8.3.0
[vagrant@ie11win7 app]$ node index.js
module.js:491
    throw err;
    ^

Error: Cannot find module '@pact-foundation/pact-mock-service-linux-x64'
    at Function.Module._resolveFilename (module.js:489:15)
    at Function.Module._load (module.js:439:25)
    at Module.require (module.js:517:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/app/node_modules/@pact-foundation/pact-mock-service/index.js:7:18)
    at Module._compile (module.js:573:30)
    at Object.Module._extensions..js (module.js:584:10)
    at Module.load (module.js:507:32)
    at tryModuleLoad (module.js:470:12)
    at Function.Module._load (module.js:462:3)
```
Which seems to be missing `npm install` step in there. Either way, I can't seem to be able to reproduce it with centos7 using your exact steps.  I didn't end up using vagrant however since it shouldn't matter just to run a few steps.

**Comment by soundstep** [comment]
I was not able to reproduce in docker either, that's why I tried vagrant.
Docker is not fully emulating an os, as the kernel is shared with your
local computer. Our DevOps never use docker to debug os related problem for
that very reason.

I will test again once I get to the office and keep you posted.

**Comment by mboudreau** [comment]
Yep, hence why I installed centos on a VM and tried it there, but still wasn't able to reproduce it.

**Comment by soundstep** [comment]
You're right, I cannot reproduce it this morning, I must have done some wrong.

I'm re-testing on the original box and I see logs like this:

```
[romuquan@jenkins test-centos-pact]$ npm install

> dtrace-provider@0.8.5 install /home/romuquan/test-centos-pact/node_modules/dtrace-provider
> node scripts/install.js


> @pact-foundation/pact-mock-service@2.1.0 postinstall /home/romuquan/test-centos-pact/node_modules/@pact-foundation/pact-mock-service
> node ./src/check-dependencies.js

[2017-08-16T08:08:32.065Z] ERROR: pact-mock-service@2.1.0/15397 on 10ft-infradev-aeuw1-jenkins-0cb3d3354739239a8: pact-node: Cannot resolve OS specific pact module. Download/install must of failed, please try again.
npm WARN test-centos-pact@1.0.0 No repository field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @pact-foundation/pact-mock-service@2.1.0 postinstall: `node ./src/check-dependencies.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @pact-foundation/pact-mock-service@2.1.0 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/romuquan/.npm/_logs/2017-08-16T08_08_32_288Z-debug.log
```

```
4649 silly postinstall @pact-foundation/pact-mock-service@2.1.0
4650 info lifecycle @pact-foundation/pact-mock-service@2.1.0~postinstall: @pact-foundation/pact-mock-service@2.1.0
4651 verbose lifecycle @pact-foundation/pact-mock-service@2.1.0~postinstall: unsafe-perm in lifecycle true
4652 verbose lifecycle @pact-foundation/pact-mock-service@2.1.0~postinstall: PATH: /home/romuquan/.n/lib/node_modules/npm/bin/node-gyp-bin:/home/romuquan/test-centos-pact/node_modules/@pact-foundation/pact-mock-service/node_modules/.bin:/home/romuquan/test-centos-pact/node_modules/.bin::/home/romuquan/.n/bin:/usr/lib64/qt-3.3/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/romuquan/.local/bin:/home/romuquan/bin
4653 verbose lifecycle @pact-foundation/pact-mock-service@2.1.0~postinstall: CWD: /home/romuquan/test-centos-pact/node_modules/@pact-foundation/pact-mock-service
4654 silly lifecycle @pact-foundation/pact-mock-service@2.1.0~postinstall: Args: [ '-c', 'node ./src/check-dependencies.js' ]
4655 silly lifecycle @pact-foundation/pact-mock-service@2.1.0~postinstall: Returned: code: 1  signal: null
4656 info lifecycle @pact-foundation/pact-mock-service@2.1.0~postinstall: Failed to exec postinstall script
4657 verbose unlock done using /home/romuquan/.npm/_locks/staging-f9e29f38d9bb6a99.lock for /home/romuquan/test-centos-pact/node_modules/.staging
4658 warn test-centos-pact@1.0.0 No repository field.
4659 verbose stack Error: @pact-foundation/pact-mock-service@2.1.0 postinstall: `node ./src/check-dependencies.js`
4659 verbose stack Exit status 1
4659 verbose stack     at EventEmitter.<anonymous> (/home/romuquan/.n/lib/node_modules/npm/lib/utils/lifecycle.js:289:16)
4659 verbose stack     at emitTwo (events.js:125:13)
4659 verbose stack     at EventEmitter.emit (events.js:213:7)
4659 verbose stack     at ChildProcess.<anonymous> (/home/romuquan/.n/lib/node_modules/npm/lib/utils/spawn.js:40:14)
4659 verbose stack     at emitTwo (events.js:125:13)
4659 verbose stack     at ChildProcess.emit (events.js:213:7)
4659 verbose stack     at maybeClose (internal/child_process.js:927:16)
4659 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
4660 verbose pkgid @pact-foundation/pact-mock-service@2.1.0
4661 verbose cwd /home/romuquan/test-centos-pact
4662 verbose Linux 3.10.0-514.26.2.el7.x86_64
4663 verbose argv "/home/romuquan/.n/bin/node" "/home/romuquan/.n/bin/npm" "install"
4664 verbose node v8.3.0
4665 verbose npm  v5.3.0
4666 error code ELIFECYCLE
4667 error errno 1
4668 error @pact-foundation/pact-mock-service@2.1.0 postinstall: `node ./src/check-dependencies.js`
4668 error Exit status 1
4669 error Failed at the @pact-foundation/pact-mock-service@2.1.0 postinstall script.
4669 error This is probably not a problem with npm. There is likely additional logging output above.
4670 verbose exit [ 1, true ]
```

**Comment by mboudreau** [comment]
Yeah, that happens when none of the optional dependencies has been
installed.  We have a postinstall script to make sure that the optional
dependency has been downloaded.  Could you run `npm install -ddd` to see
all logs of what's going on behind the scenes?

**Comment by soundstep** [comment]
The problem is, it is not all time, so it is kind of hard to reproduce.

I was always wondering why it says ` pact-mock-service@2.1.0/9511 on 10ft-infradev-aeuw1-jenkins-0cb3d3354739239a8: Platform specific dependency '@pact-foundation/pact-mock-service-darwin' installed successful` on a linux system.

Silly question but... could this be related to network problem somehow?

When I see:

```
[2017-08-16T08:08:32.065Z] ERROR: pact-mock-service@2.1.0/15397 on 10ft-infradev-aeuw1-jenkins-0cb3d3354739239a8: pact-node: Cannot resolve OS specific pact module. Download/install must of failed, please try again.
```

Could it be a download that fails for some reason?

Still investigating but right now it is working, I'll try a bit later, I know the problem is still there.

**Comment by mboudreau** [comment]
yeah, I did see that today as well.  I'm not too sure what's going on
because when I actually look at the folders, there's nothing there for
'darwin' only 'linux-x64', however, this chat has sparked an idea in my
head that I'll have to check out tomorrow morning.

**Comment by mefellows** [comment]
I suppose it could be @soundstep . I've just spun up a clean centos ec2 64 bit instance with your steps:

```
curl --silent --location https://rpm.nodesource.com/setup_8.x | sudo bash -
sudo yum -y install nodejs
npm install pact
./node_modules/\@pact-foundation/pact-mock-service/node_modules/.bin/pact-mock-service
```

Worked a charm.

I'm wondering if as you say, it could be network related or something else. Are you running this inside of a corporate proxy or something similar? It might be worth attempting an install of a different version of pact-node to  work around any potential caching, or even running outside of your work network just to rule it out.

FWIW I recently had issues behind a corporate proxy and had to make a few adjustments to a dockerfile to make it work:

```
npm config set maxsockets 5
npm config set registry http://registry.npmjs.org/
npm config set strict-ssl false
```

There's a post somewhere that mentions npm seems to be a load testing tool for routers...

**Comment by soundstep** [comment]
Right interesting @mefellows !
I'm trying to gather a bit more logs.

Do you guys mind check that one: https://gist.github.com/soundstep/68e47a96f784abb08104aacba4d32089

It looks like it has installed the pact=mock-service for darwin, while this is happening on the Centos 7 machine.

Am I reading the logs wrong?

**Comment by mboudreau** [comment]
@soundstep no, you're definitely on the right track, those logs are saying that it is downloading the darwin version, which is weird.  I'll investigate further, but I'm not sure why you're seeing it consistently while I can run pact without a hitch on the latest centos7.

**Comment by soundstep** [comment]
Actually, it is not consistently, it is like 50/50. Sometimes it works, sometimes it doesn't.
I think I've noticed that starting from a clean workspace (no node_modules folder), it will always fail. Then re-running a job on the same workspace (with a previous node_modules), it sometimes work.

**Comment by soundstep** [comment]
Note that we are also using a private npm registry on artifactory, I hope it has nothing to do with that and some sort of caching. Actually, I can use the normal npm registry on this project, I will disable it just to make sure!

**Comment by mboudreau** [comment]
@soundstep hm, artifactory caching might be the cause.  Kinda hard to reproduce on my end ;)

**Comment by soundstep** [comment]
@mboudreau nevermind, it is not used on this project. I'll double check the npmrc file is correct, just to make sure.

**Comment by soundstep** [comment]
I think I finally understood what's happening: the package-lock.json file!

We are developing on MacOS computer, and sadly, pushing the package-lock.json file to our repo.
This makes NPM install the darwin version of the pact-mock-service on the linux machine, as it has been stored in the lock file.

Here is what show in the package-lock:

```
"dependencies": {
    "@pact-foundation/pact-mock-service": {
      "version": "0.10.2",
      "resolved": "http://itvrepos.jfrog.io/itvrepos/api/npm/npm-itv/@pact-foundation/pact-mock-service/-/pact-mock-service-0.10.2.tgz",
      "integrity": "sha1-LukJYpZNyk86HEgw4Fd7/GGtgKk=",
      "requires": {
        "@pact-foundation/pact-mock-service-darwin": "0.10.2",
        "bunyan": "1.8.12",
        "bunyan-prettystream": "0.1.3"
      }
    },
    "@pact-foundation/pact-mock-service-darwin": {
      "version": "0.10.2",
      "resolved": "http://itvrepos.jfrog.io/itvrepos/api/npm/npm-itv/@pact-foundation/pact-mock-service-darwin/-/pact-mock-service-darwin-0.10.2.tgz",
      "integrity": "sha1-Wphp1pr8Z8hbkDOEmGmRJBuHXok=",
      "optional": true
    },
```

So there is no bug in pact-node, which is great news but I think it is a little weakness with the `optionalDependencies` field.

We are used to push Gemfile.lock file to our repo because it is a good thing, so we thought it would be the same thing for node. I think I will now add the package-lock.json file in our `.gitignore`.

I will link this issue to npm just so they think about this issue.

Bottom line, thank you guys for the help, you were so responsive, you're awesome ;)

**Comment by mboudreau** [comment]
@soundstep No way! Of course I forgot about that since it's a new thing for npm 5. I definitely recommend you add the package.lock file to your .gitignore and delete it from the repo as not only is this going to cause issues with your build server, but any person trying to develop on your project that's not on a mac.

Thanks for bearing with us and giving us an abundant amount of info to help us debug. Don't hesitate to bring up more bugs if you're always going to be this thorough ;)

**Comment by mefellows** [comment]
Great spot!!

Hmm this is another pain point. It's good practice to check in the
package-lock.json so recommending against this approach is not something
I'm a fan of. IMO we should consider removing the optional dependencies.

I'd rather see us pull down the correct binaries to install during the post
install phase instead.

We could then also get rid of the other packages (mock service and
verifier) , something that's a frustrating aspect of our release process.

We then just have two components - pact-node (engineer wrapper) and pact
(the dsl).

Thoughts?

**Comment by mboudreau** [comment]
The only other possibility is to load all OS specific dependencies, which
is then required based on the runtime, but this makes the install process
heavy since we're now installing all binaries.  Postinstall process won't
work since it isn't triggered on npm update, which can cause some
versioning issues.

**Comment by soundstep** [comment]
What I would recommend is that you install/download/build the os specific stuff during an `npm rebuild`, which happen during an `npm install` (npm build).

https://docs.npmjs.com/cli/rebuild

That's what lib-sass is doing for example. if you switch machine or node version, they log a message like "Please execute `npm rebuild lib-sass`".

Another example, it is considered a good practice when building nodejs micro-services using RPM, to have the node_modules folder included in the RPM (as it should be self-contained). So people usually trigger a `npm rebuild` when they update their services on linux, as the RPM might be built on another type of machine. Like this guy there: https://github.com/bbc/speculate#install-your-dependencies-first

Point is, if you execute `npm rebuild`, it should re-download or re-build anything that is os-specific.
I think it makes sense.

**Comment by mboudreau** [comment]
Good points, I'll have to look into that.

**Comment by soundstep** [comment]
We heavily use pact in production, so I'll definitely be there if we find issues ;)
We fixed a few in the ruby pact-mock-service, we also made https://github.com/ITV/scala-pact

**Comment by mefellows** [comment]
@soundstep I was on a mobile before, but I was essentially trying to say the same thing. We don't actually go through the same compile as saas does however (which does a native compile with `node-gyp`), but it is analogous.

I think it's acceptable to force an npm reinstall for this if an update is required. If [angular cli](https://github.com/angular/angular-cli#updating-angular-cli) can get away with it -

> rm -rf node_modules dist
> rm -r -fo node_modules,dist in Windows PowerShell
> npm install --save-dev @angular/cli@latest
> npm install


- then surely we can too!

**Comment by mefellows** [comment]
Also, we love the work you guys (including Dave) have done at ITV!

**Comment by soundstep** [comment]
@mefellows I don't think having to re-do an `npm install` is the solution, that's why I specifically gave you the other example with the node micro-service bundle in an RPM. They are supposed to be ready-to-use and self-contained.

You would never do an npm install in a production environment, that's not how softwares should be deployed, especially not relying on network or npm registry. Ideally, it should rely on nothing so that your deployment is the safest possible.

What I mean is, you should be able to deploy a node app without accessing to the network. The only thing you need to do is a `npm rebuild` to satisfy os-dependencies, things that can't be done outside of the actual environment.

You could argue that pact-node is a build tool and wouldn't be used that way, that's fine. But I would still do it the right way, who knows how pact-node should be used and you should actually not ncare about that.

What would be the problem in getting the right pact-mock-service in a `preinstall` script for example?

If you run a `npm rebuild`, the preinstall script and others will be re-executed.

```
"scripts": {
    "preinstall": "get my correct pact-mock-service dependency"
```

Hope that helps!

**Comment by mefellows** [comment]
Agree with all of those points. Provided the install hooks fire as it says
on the label this should be easy, there should be no problem..

We just need to try it out and test it is all I'm saying.

On Aug 18, 2017 7:20 PM, "Romuald Quantin" <notifications@github.com> wrote:

@mefellows <https://github.com/mefellows> I don't think having to re-do an npm
install is the solution, that's why I specifically gave you the other
example with the node micro-service bundle in an RPM. They are supposed to
be ready-to-use and self-contained.

You would never do an npm install in a production environment, that's not
how softwares should be deployed, especially not relying on network or npm
registry. Ideally, it should rely on nothing so that your deployment is the
safest possible.

What I mean is, you should be able to deploy a node app without accessing
to the network. The only thing you need to do is a npm rebuild to satisfy
os-dependencies, things that can't be done outside of the actual
environment.

You could argue that pact-node is a build tool and wouldn't be used that
way, that's fine. But I would still do it the right way, who knows how
pact-node should be used and you should actually not ncare about that.

What would be the problem in getting the right pact-mock-service in a
preinstall script for example?

If you run a npm rebuild, the preinstall script and others will be
re-executed.

"scripts": {
    "preinstall": "get my correct pact-mock-service dependency"

Hope that helps!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/pact-foundation/pact-node/issues/42#issuecomment-323304271>,
or mute the thread

**Comment by soundstep** [comment]
@mefellows Yep, just checked it, the preinstall script is executed on `npm rebuild`, thus `npm intall`.
You should all fine with that!
Cheers.

**Comment by mefellows** [comment]
Great! One thing to note is that we will require network access in this
phase to pull down the OS specific bundle.

But whichever way you look at it, it's better than the current situation.

On Aug 18, 2017 7:46 PM, "Romuald Quantin" <notifications@github.com> wrote:

@mefellows <https://github.com/mefellows> Yep, just checked it, the
preinstall script is executed on npm rebuild, thus npm intall.
You should all fine with that!
Cheers.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/pact-foundation/pact-node/issues/42#issuecomment-323310338>,
or mute the thread

**Comment by soundstep** [comment]
Yeah good point. Can't really help with that I guess. 
And I guess the pact-mock-service are big enough so that you can't include them all?

**Comment by mefellows** [comment]
Exactly. Longer term we will probably be able to compile in this step
though.

On Aug 18, 2017 8:13 PM, "Romuald Quantin" <notifications@github.com> wrote:

Yeah good point. Can't really help with that I guess.
And I guess the pact-mock-service are big enough so that you can't include
them all?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/pact-foundation/pact-node/issues/42#issuecomment-323316137>,
or mute the thread

**Comment by travi** [comment]
finally got one of my projects that uses pact-node upgraded to node 8 and ran into this problem. i do not agree with the recommendation to ignore the lockfile since the whole point of the lockfile is to keep package versions consistent across machines.

i tried to run the `npm rebuild` on travis, but (i assume) since i was still versioning the lockfile, i got 
```
pact-node: Cannot resolve OS specific pact module. Download/install must of failed, please try again.
```

i was able to get the travis build to work by running `npm install @pact-foundation/pact-mock-service-linux-x64` as a `before_script` step on travis. this works for my team since we are all developing on osx, but i'm not sure how to solve this with the current approach for teams that have developers across OSs.

it looks like npm 5 doesnt support this type of approach very well yet, but it looks like you've got things set up correctly for when it does, at least from what i understand. might be worth keeping track of issues like:
* https://github.com/npm/npm/issues/17519

**Comment by mefellows** [comment]
Thanks @travi, there's a bit of work for us to get to our optimal solution, and your workaround seems appropriate for now. I'm watching both of those issues to see if our recommendation changes, but in any case I feel the optional dependencies approach needs to be revised.

**Comment by TimothyJones** [comment]
I'm also hitting this problem with a dockerised build. I'll try @travi's workaround.

**Comment by mefellows** [comment]
Thanks Tim - we're working on it, we promise!

**Comment by mefellows** [comment]
FYI see https://github.com/pact-foundation/pact-js/issues/91 (on our [project board](https://github.com/pact-foundation/pact-js/projects/2) for a proposal with options related to this issue and beyond. Thoughts welcome!

**Comment by mboudreau** [comment]
@mefellows @travi @soundstep 

After some though, the rebuild option simply will not work for our purpose as we need npm's internals to install our binary and packages properly using npm link and should be resolvable using require.resolve.  If we are to use npm rebuild, many of these things won't happen.  Package structures can get screwy between npm versions, hence the need for require.resolve.  I had tried the rebuild option (originally just a postinstall with npm install in it) when I originally made this package, but problems arose with the new flat structure of npm and that it might revert back to a hierarchical one depending on context.  The other issue is that it would actually download and install the package every time `npm install` was ran, meaning it's not a normal one-off situation.

Using optionalDependencies was the best solution from a developer experience standpoint, and now npm seems to have made our lives difficult again with changing how it works under the covers with package-lock.json file not support a core feature of npm.

As far as I'm concerned, this is not a pact-node issue, but an npm issue: https://github.com/npm/npm/issues/18135

Only way to properly fix this would be to include all binaries of all OS in a single package, then have a simple script that checks the user's OS and use the correct binary when pact-node is used.  Can't say I like the idea of having pact-node be ~28mb for each download, but if we have no choice...

**Comment by travi** [comment]
@mboudreau i agree 100%. i think you had the dependencies configured correctly from the start here, but are facing what i consider a clear npm bug. i'm really surprised that npm/npm#18135 doesnt seem to have much priority behind it yet.

**Comment by mboudreau** [comment]
@travi yeah, sadly.  I might have to use my recommendation at the end up loading all binaries and doing runtime check as to while one to grab :/

@mefellows Thoughts on this?

**Comment by mboudreau** [comment]
@travi @soundstep @mefellows Good news everyone.

Pact standalone now packages all binaries in a single dependency that's not OS specific anymore.  This shitty issue with npm optional dependencies and package-lock.json should now be fixed for pact-node in version 5.1.2.  Please try it out and report back to see if it works for your situation.

Very sorry about this whole ordeal.  Cheers.

**Comment by mefellows** [comment]
Thanks so much @mboudreau for all your work on a _bunch_ of stuff this past few days - I encourage anyone to look at the commit history to see this dedication.

**Comment by mboudreau** [comment]
I'm going to close this ticket as I haven't heard back from anyone, but this should be fixed by now.  If someone doesn't agree, I'll reopen.

**Comment by TimothyJones** [comment]
@mefellows is publishing a new version of `pact-js` that uses this version - I'll test it for our case and confirm that our workaround isn't necessary any more.

**Comment by mefellows** [comment]
FYI pact `4.0.x` was released today, please let us know how you go!

**Comment by soundstep** [comment]
Sorry @mboudreau and @mefellows, I was silent because I didn't have a chance to test!
But thanks everyone for the hard work!

**Comment by mefellows** [comment]
No worries @soundstep, whenever you get the chance mate.

**Comment by bbarke** [comment]
I ran into this same issue - using 4.12.
I upgraded to 5.2.0 and the issue was resolved.
Thankyou

**Comment by TimothyJones** [comment]
Which reminds me - we tested this, and it works perfectly. Thanks, sorry for forgetting to update!

**Comment by mboudreau** [comment]
@soundstep Just as a closing comment, we've now merged the pact-standalone with in pact-node, starting at version 6.12.0.  Only the OS specific version is downloaded on postinstall and not all binaries as to remove the need for a weighty package that has all operating systems.

Give it a whirl :D
